src/flatten_ok.rs

```diff
@@ -0,0 +1,61 @@
+pub fn flatten_ok<I, T, E>(iter: I) -> FlattenOk<I, T, E>
+where
+    I: Iterator<Item = Result<T, E>>,
+    T: IntoIterator,
+{
+    FlattenOk { iter, inner: None }
+}
+
+/// An iterator adaptor that flattens `Result::Ok` values and
+/// allows `Result::Err` values through unchanged.
+///
+/// See [`.flatten_ok()`](crate::Itertools::flatten_ok) for more information.
+#[must_use = "iterator adaptors are lazy and do nothing unless consumed"]
+pub struct FlattenOk<I, T, E>
+where
+    I: Iterator<Item = Result<T, E>>,
+    T: IntoIterator,
+{
+    iter: I,
+    inner: Option<T::IntoIter>,
+}
+
+impl<I, T, E> Iterator for FlattenOk<I, T, E>
+where
+    I: Iterator<Item = Result<T, E>>,
+    T: IntoIterator,
+{
+    type Item = Result<T::Item, E>;
+
+    fn next(&mut self) -> Option<Self::Item> {
+        loop {
+            if let Some(inner) = &mut self.inner {
+                if let Some(item) = inner.next() {
+                    return Some(Ok(item));
+                } else {
+                    self.inner = None;
+                }
+            } else {
```

Comment:
Do you really need this `else`? The only case where it isn't executed is if `self.inner` is `Some(_)` and `inner.next()` is also `Some(_)`, but in that case you have an early `return`. I would execute the contents of the `else` in any case, both if the `if let` succeeds or not.

Reply:
@SkiFire13 Your point is sound. In fact, we shouldn't even need to assign it to `None` in that case as it will just automatically fall through. That might hurt performance in the case where `next()` is called on an empty iterator repeatedly, but I don't think that actually matters. So yes, let me make that change.

Reply:
I think setting it to `None` might have a reason if you consider consistency in case `next` is called after the iterator ended. Should it try to advance the last inner iterator? Or the outer iterator? Currently the stdlib fuse both of them, but I guess it would be reasonable if you advance the outer iterator (in that case you need to set the `self.inner` to `None`)

Reply:
I realize now that I should not change setting the inner to `None` as it might incur a performance penalty when hitting several `Err` values in a row.